### PR TITLE
Migrated from Talons to SparkMaxes

### PR DESCRIPTION
**Pull request by Nins97**
This is kind of terrible and needs some cleanup. I'm not even sure if it works :(

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 1]
```diff
@@ -73,8 +75,8 @@ private enum SwerveModuleControlMode {
     private final Alert inDeadModeAlert;
 
     private final TelemetryTalonFX driveMotor;
-    private final TelemetryTalonFX steerMotor;
-    private final CANCoder absoluteSteerEncoder;
+    private final TelemetryCANSparkMax steerMotor;
+    private final RelativeEncoder absoluteSteerEncoder;
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:79]: The CANCoder is still needed as the CANCoder is the absolute encoder. You'll need another `steerEncoder` object

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 2]
```diff
@@ -89,6 +91,8 @@ private enum SwerveModuleControlMode {
     private final TunablePIDGains steerPositionPIDGains;
     private SimpleMotorFeedforward driveMotorFF;
 
+    SparkMaxPIDController steerController;
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:94]:
```suggestion
    private final SparkMaxPIDController steerController;
```

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 3]
```diff
@@ -139,18 +144,14 @@ public SwerveModule(SwerveModuleConfiguration config, boolean tuningMode) {
...
-        this.steerMotor = new TelemetryTalonFX(
-                config.steerMotorPort(),
-                tableName + "steerMotor",
-                config.sharedConfiguration().canBus(),
-                tuningMode);
-        configSteerMotor(config);
-
+        this.steerMotor = new TelemetryCANSparkMax(
+                config.steerMotorPort(), CANSparkMax.MotorType.kBrushless, tableName + "steerMotor", tuningMode);
+        //        configSteerMotor(config);
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:153]: Uncomment out

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 4]
```diff
@@ -203,54 +204,81 @@ private void configDriveMotor(SwerveModuleConfiguration config) {
...
         ConfigTimeout configTimeout = new ConfigTimeout(MiscConstants.CONFIGURATION_TIMEOUT_SECONDS);
         boolean faultInitializing;
+
         do {
-            faultInitializing = checkCTREError(steerMotor.configFactoryDefault());
+            faultInitializing = checkRevError(steerMotor.restoreFactoryDefaults());
 
-            TalonFXConfiguration motorConfiguration = new TalonFXConfiguration();
-            applyCommonMotorConfiguration(motorConfiguration, config);
+            //             motorConfiguration = new TelemetryCANSparkMax.();
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:216]: remove commented out code

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 5]
```diff
@@ -203,54 +204,81 @@ private void configDriveMotor(SwerveModuleConfiguration config) {
...
+
+            //            absoluteSteerEncoder.setInverted(
+            //                    config.steerMotorInverted() == true ? absoluteSteerEncoder.setInverted(true) :
+            // absoluteSteerEncoder.setInverted(false));
+
+            if (config.steerMotorInverted()) {
+                absoluteSteerEncoder.setInverted(true);
+            } else {
+                absoluteSteerEncoder.setInverted(false);
+            }
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:261]:
```suggestion
            absoluteSteerEncoder.setInverted(config.steerMotorInverted());
```

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 6]
```diff
@@ -203,54 +204,81 @@ private void configDriveMotor(SwerveModuleConfiguration config) {
...
+            //            config.sharedConfiguration().steerPositionPIDGains().setSlot(motorConfiguration.slot0);
             //            motorConfiguration.slot0.allowableClosedloopError =
             //                    config.sharedConfiguration().allowableSteerErrorRadians() /
             // steerMotorConversionFactorPosition;
-            motorConfiguration.slot0.closedLoopPeakOutput = config.sharedConfiguration()
-                            .maxSteerVoltage()
-                    / config.sharedConfiguration().nominalVoltage();
 
-            faultInitializing |= checkCTREError(steerMotor.configAllSettings(motorConfiguration, CAN_TIMEOUT_MS));
+            // TODO: I have no clue if this works
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:240]: I'm not sure either, the parameters are min and max, but I don't know the units

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 7]
```diff
@@ -295,17 +325,17 @@ private void reportError(String message) {
...
+    //        if (RobotBase.isReal() && steerMotor.getL()) {
+    //            reportError("Steer motor reset occurred");
+    //            setToAbsolute = false;
+    //            resetSteerToAbsolute();
+    //        }
+    //    }
 
     private void checkForDriveMotorReset() {
-        if (RobotBase.isReal() && steerMotor.hasResetOccurred()) {
+        if (RobotBase.isReal() && driveMotor.hasResetOccurred()) {
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:338]: nice catch

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 8]
```diff
@@ -331,15 +361,17 @@ public void resetSteerToAbsolute(double timeout) {
         boolean gotAbsolutePosition = false;
         do {
             absolutePosition = getAbsoluteRadians();
-            if (!checkCTREError(absoluteSteerEncoder.getLastError()) || timeout == 0.0) {
+            if (steerMotor.getLastError() == null || timeout == 0.0) {
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:364]: this should stay how it was

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 9]
```diff
@@ -479,20 +511,20 @@ public void setRawVoltage(double driveVolts, double steerVolts) {
         controlModeEntry.append(SwerveModuleControlMode.RAW_VOLTAGE.logValue);
 
         driveMotor.set(TalonFXControlMode.PercentOutput, driveVolts / nominalVoltage);
-        steerMotor.set(TalonFXControlMode.PercentOutput, steerVolts / nominalVoltage);
+        steerMotor.setVoltage(steerVolts / nominalVoltage);
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:514]: don't divide by nominalVoltage if you are setting voltage directly

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 10]
```diff
@@ -479,20 +511,20 @@ public void setRawVoltage(double driveVolts, double steerVolts) {
...
 
         if (activeSteer) {
-            steerMotor.set(
-                    TalonFXControlMode.Position,
+            absoluteSteerEncoder.setPosition(
                     RaiderMathUtils.calculateContinuousInputSetpoint(getSteerAngleRadiansNoWrap(), targetAngleRadians)
                             / steerMotorConversionFactorPosition);
         } else {
-            steerMotor.neutralOutput();
+            steerMotor.setIdleMode(CANSparkMax.IdleMode.kBrake);
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:526]:
```suggestion
            steerMotor.setVoltage(0.0);
```

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 11]
```diff
@@ -526,7 +558,10 @@ private void checkAndUpdateGains() {
         if (steerPositionPIDGains.hasChanged()) {
             SlotConfiguration newSlotConfig = new SlotConfiguration();
             steerPositionPIDGains.setSlot(newSlotConfig);
-            steerMotor.configureSlot(newSlotConfig);
+            steerController = steerMotor.getPIDController();
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:561]: I don't think you need to do this as it is already the correct reference

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 12]
```diff
@@ -526,7 +558,10 @@ private void checkAndUpdateGains() {
         if (steerPositionPIDGains.hasChanged()) {
             SlotConfiguration newSlotConfig = new SlotConfiguration();
             steerPositionPIDGains.setSlot(newSlotConfig);
```
Comment (ohowe1) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:560]: you don't need this as SlotConfiguration is phoenix object

**Review by ohowe1 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java (hunk 11 above)
Comment (Nins97) [reply src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:561]: nuh uh

**Review by ohowe1 — APPROVED** [review]

**Pull request merged**